File: agent-api/app/services/files/deliverable.py
# -*- coding: utf-8 -*-
"""交付物判据（2026-07-27 用户拍板）——什么算「交给用户的东西」，什么只是过程文件。

## 为什么要这条闸

统一文件系统落地后「写进 /workspace/files 就是已保存」，代价是模型的**过程文件也一起进了
用户的文件区**：做一份 PPT 时先写 `build.py` 生成脚本，用户就在对话里收到一张
`workspace_tmp_build.py` 的产物卡（PY · 15.1 KB · 已保存到我的文件），真正的 pptx 混在里面。
`workspace_sync._INTERMEDIATE_EXTS` 那条黑名单只管 bash 回写这一条路径，`write_file` /
`create_file` / `execute_in_sandbox` 三条路都绕过它。

用户拍板：**产物卡与「我的文件」都只展示文档类产物**（word / pdf / excel / ppt / md 之类，
任务模式的《任务计划.md》天然在内），其余（.py 脚本、.json 中间数据、日志……）照常落库供模型
下一步继续用，但不作为交付物露出。

## 白名单而不是黑名单

`workspace_sync` 当初选黑名单的理由是「交付格式是开放集合，白名单会静默吞掉没预料到的格式」。
用户 2026-07-27 明确反转了这个取舍：宁可漏掉少数冷门格式，也不要文件区被过程文件淹没。
**网页（.html/.htm）保留**——平台本来就能行内预览它，用户特意点名要留（"我们之前不也是有
html 可以展示吗"）。图片保留同理：execute_in_sandbox 那条路已经把「嵌进文档的中间图」滤掉了，能走到
这里的图片就是用户真要的那张。

## 两条轴：格式 + 来源

光看格式不够。真机第二例：用户只说「扫描一下这个 GitHub 仓库的代码」，模型用 `download_url`
逐个取回 16 个源文件，文件区一次多出 `main.tsx` / `package.json` / `docker-compose.yml` /
`README.md` / `requirements.txt`……**后两个是白名单里的格式**，只按后缀判还是会露出来。

所以 `download_url` 落地的文件单独标 `source="material"`（取来的原材料）：它是**输入**，
不是本轮的产出。代价是「帮我下载这个 PDF」时那份 PDF 也不出卡——用户仍能在「我的文件」
里打开「显示全部文件」看到它（清单默认藏、不删除，见 routers/files.py 的 `show_all`）。

## 「我的文件」页上传永远可见；对话框附件不是

`source == "uploaded"` 一律判可交付：用户在「我的文件」页主动上传的 .py/.zip 是
**他自己的文件**，平台没资格因为后缀不在白名单里就藏掉。

对话框 / composer 放下的图片、PPT、粘贴文本走 `source=workspace`，那是本轮素材，
不是产物，清单（含「显示全部文件」）不露出。成品 PPT 仍只经发布进「我的文件」。
"""
from __future__ import annotations

import os
from typing import Optional

#: `download_url` 取回的外部字节。既不是用户传的、也不是本轮产出的，属于「材料」。
MATERIAL_SOURCE = "material"
#: PPTD 工程跨 Run 检查点。不是交付物，也不进模型 list_files / 沙箱 files 镜像。
#: 必须 ≤16 字符：`agent_user_file.source` / `agent_user_file_version.source` 是 VARCHAR(16)。
#: 真机 `staging_checkpoint`（19）INSERT 报 Data too long，检查点打包成功却落不了库。
STAGING_CHECKPOINT_SOURCE = "staging_ckpt"
#: 工作区对象若误写入「我的文件」表，也不得当交付物露出。
WORKSPACE_SOURCE = "workspace"
_STAGING_CHECKPOINT_SOURCES = frozenset({
    STAGING_CHECKPOINT_SOURCE,
    "staging_checkpoint",  # 旧名：兼容未写入成功前的测试与注释
    WORKSPACE_SOURCE,
})
assert len(STAGING_CHECKPOINT_SOURCE) <= 16


def is_staging_checkpoint_source(source: Optional[str] = None) -> bool:
    return str(source or "") in _STAGING_CHECKPOINT_SOURCES

# 文档类：用户口径的「交付产物」。任务模式投影的《任务计划.md》落在 .md 上。
_DOC_EXTS = frozenset({
    ".docx", ".doc", ".pptx", ".ppt", ".xlsx", ".xls", ".pdf",
    ".md", ".markdown", ".txt", ".csv", ".tsv",
})
# 网页：平台有行内预览 + 全屏查看器，是能直接交给用户看的东西（用户点名保留）
_WEB_EXTS = frozenset({".html", ".htm"})
# 图片：能走到交付层的图片已是最终产物（中间素材在 execute_in_sandbox harvest 阶段就被滤掉了）
_IMAGE_EXTS = frozenset({".png", ".jpg", ".jpeg", ".gif", ".bmp", ".webp", ".svg"})

DELIVERABLE_EXTS = _DOC_EXTS | _WEB_EXTS | _IMAGE_EXTS

#: 编辑器的内部伴生文件：**不是交付物，但清单里必须带着**。
#: `<名>.slides.json` 是幻灯片编辑源；`<名>.research.md` 是研究报告 Markdown 源。
_COMPANION_SUFFIXES = (".slides.json", ".research.md")
#: 办公批量交付打包名。其它 generated zip 仍不是产物卡（见 test_generated_ppt_source_bundle）。
_BATCH_DELIVERABLE_ARCHIVES = frozenset({"batch-deliverables.zip"})


def is_editor_companion(filename: str) -> bool:
    return str(filename or "").strip().lower().endswith(_COMPANION_SUFFIXES)


def is_deliverable(filename: str, source: Optional[str] = "generated") -> bool:
    """这个文件该不该作为交付产物露给用户（产物卡 / 「我的文件」清单 / 交付审查）。

    判据只有三条，故意不看内容也不看体积——判据要能一句话说清楚，否则用户下次看到
    某个文件没出现时无从解释。
    """
    if is_editor_companion(filename):
        return False
    src = str(source or "")
    if src == MATERIAL_SOURCE or is_staging_checkpoint_source(src):
        return False  # 材料与 PPTD 检查点都不是交给用户的产物
    if src not in ("generated", "research", ""):
        return True  # 上传件 / 恢复的历史版本：用户自己的东西，一律可见
    name = os.path.basename(str(filename or "").strip().lower())
    if name in _BATCH_DELIVERABLE_ARCHIVES:
        return True
    ext = os.path.splitext(str(filename or "").strip().lower())[1]
    return ext in DELIVERABLE_EXTS


def filter_rows(rows) -> list:
    """按行过滤落库结果 dict 列表（`{"filename": ..., "source": ...}` 形状）。

    行里若已带 `deliverable`（`_row_to_dict` 算好的）就直接用，避免同一判据算两遍、
    也让「历史行没有这个字段」时能按文件名兜底。
    """
    out = []
    for row in rows or []:
        if not isinstance(row, dict):
            continue
        flag = row.get("deliverable")
        if flag is None:
            flag = is_deliverable(
                str(row.get("filename") or row.get("name") or ""),
                row.get("source"),
            )
        if flag:
            out.append(row)
    return out

File: src/views/peopleCenter/composables/deliverable.ts
/**
 * 交付物判据（2026-07-27 用户拍板）——哪些文件该作为「产物」露给用户。
 *
 * 与后端 `agent-api/app/services/files/deliverable.py` 的 `DELIVERABLE_EXTS` 是**同一份清单**，
 * 改一边必须改另一边（下方 `DELIVERABLE_EXTS` 有对应的 pytest/jest 各自守着）。
 *
 * 为什么前端还要再判一次而不是只信后端下发的 `deliverable` 字段：历史消息回放时，
 * 事件里的文件行是**当初**序列化的，没有这个字段。只认字段的话老对话里的 build.py 卡片
 * 会一直挂着。所以字段优先、按后缀兜底。
 *
 * 产物卡里的文件必然是本轮生成的（source=generated），因此这里不需要像后端那样为
 * 「用户自己上传的 .py」开豁免。
 */

/** 文档类：用户口径的交付产物（任务模式的《任务计划.md》落在 .md 上） */
const DOC_EXTS = ['.docx', '.doc', '.pptx', '.ppt', '.xlsx', '.xls', '.pdf', '.md', '.markdown', '.txt', '.csv', '.tsv'];
/** 网页：平台自己就能预览（用户明确点名保留） */
const WEB_EXTS = ['.html', '.htm'];
/** 图片：能走到交付层的已是最终产物（嵌进文档的中间图在后端 harvest 阶段就滤掉了） */
const IMAGE_EXTS = ['.png', '.jpg', '.jpeg', '.gif', '.bmp', '.webp', '.svg'];

export const DELIVERABLE_EXTS = [...DOC_EXTS, ...WEB_EXTS, ...IMAGE_EXTS];

export function isDeliverableName(filename: string): boolean {
  const name = String(filename || '').trim().toLowerCase();
  const dot = name.lastIndexOf('.');
  if (dot <= 0) return false; // 无扩展名 / 纯 dotfile：不是文档
  return DELIVERABLE_EXTS.includes(name.slice(dot));
}

export function isDeliverableFile(file: { filename?: string; deliverable?: boolean } | null | undefined): boolean {
  if (!file) return false;
  if (typeof file.deliverable === 'boolean') return file.deliverable;
  return isDeliverableName(file.filename || '');
}

/** 「我的文件」页打开时就带上「显示全部文件」的查询参数（MyFilesTab 读它初始化 showAll）。 */
export const MY_FILES_SHOW_ALL_QUERY = 'all=1';

/**
 * 时间线上「已下载 报告.pdf」「已写入 build.py」那颗按钮该跳到哪里。
 *
 * 「我的文件」默认清单是 `deliverables_only`：`download_url` 取回的材料（source=material，
 * **不看后缀**）和 .py/.json/.zip 这类过程文件都被滤掉。按钮上明明写着文件名、点下去却落进
 * 一片找不到它的列表——这比干脆不给按钮更让人confused。所以这类目标带上 `all=1` 落地，
 * 页面直接开着「显示全部文件」。
 *
 * 交付物照旧跳干净的默认视图：默认视图之所以克制，就是为了别让产物淹在过程文件里。
 *
 * 判不准时**偏向 all=1**：这里只看得到文件名和工具名，看不到 source。用户自己上传的 .py
 * 在后端是可交付的（上传件永远可见），按后缀会被这里判成非交付物 → 跳到全量视图。
 * 全量是默认视图的超集，代价只是列表吵一点；反过来判错就是跳到一个空列表。
 */
export function myFilesRouteFor(step: { name?: string; target?: string }): string {
  const visibleByDefault = step.name !== 'download_url' && isDeliverableName(step.target || '');
  return visibleByDefault ? '/center/files' : `/center/files?${MY_FILES_SHOW_ALL_QUERY}`;
}

type FileLike = { filename: string; deliverable?: boolean };

/**
 * 会真的出现在产物卡里的文件。
 * ① 交付物白名单（上面的判据）；
 * ② `.slides.json` 是幻灯片编辑器的内部编辑源（2026-07-21 用户拍板），数据留在
 *    generatedFiles 里供配对/「编辑」按钮用，但不作为文件卡露出。
 * ③ `.research.md` 是研究报告的 Markdown 伴生源，下载菜单再取，不单独出卡。
 */
const COMPANION_NAME_RE = /\.(slides\.json|research\.md)$/i;

export function isCompanionFile(filename: string | null | undefined): boolean {
  return COMPANION_NAME_RE.test(String(filename || ''));
}

export function visibleDeliverables<T extends FileLike>(files: T[] | null | undefined): T[] {
  return (files || []).filter((f) => !isCompanionFile(f.filename) && isDeliverableFile(f));
}

/**
 * 本轮是否交付了实质内容——**必须与产物卡同口径**，所以两者共用 visibleDeliverables。
 *
 * 2026-07-27 事故：判据写成「generatedFiles 非空」，而产物卡按交付物白名单筛。模型先
 * write_file 写了 build.py、本轮随后失败且正文为空时，两个口径打架——判定「已交付」于是
 * 隐掉错误横幅、执行头按正常终态显示，产物卡那边又把 build.py 滤掉，用户得到一条没有正文、
 * 没有卡片、也没有报错的空气泡。
 */
export function hasDeliveredContent(
  content: string | null | undefined,
  files: FileLike[] | null | undefined,
): boolean {
  return Boolean(String(content || '').trim() || visibleDeliverables(files).length);
}
